File: tests/test_v7_user_ws.py
from __future__ import annotations

import shutil
import subprocess
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
PROGRAM = r'''
#include "pm/v7_user_ws.hpp"
#include <array>
#include <cassert>
#include <iostream>
using namespace pm::v7::user_ws;
int main(){
 auto o=decode(R"({"event_type":"order","id":"0xORDER","market":"0xCOND","asset_id":"TOKEN","side":"SELL","price":"0.57","original_size":"10","size_matched":"0","type":"PLACEMENT","status":"LIVE"})",100);
 assert(o.recognized && !o.invalid && o.event.kind==EventKind::Order);
 assert(o.event.order_type==OrderEventType::Placement && !o.event.buy_side);

 auto t=decode(R"({"event_type":"trade","id":"trade-uuid","taker_order_id":"taker-123","market":"0xCOND","asset_id":"TOKEN","side":"BUY","size":"10","price":"0.57","status":"MATCHED","type":"TRADE","trader_side":"taker","maker_orders":[{"order_id":"maker-a","owner":"owner","matched_amount":"2.5","price":"0.57","asset_id":"TOKEN","side":"SELL"},{"order_id":"maker-b","owner":"owner","matched_amount":"7.5","price":"0.57","asset_id":"TOKEN","side":"SELL"}]})",120);
 assert(t.recognized && !t.invalid && t.event.kind==EventKind::Trade);
 assert(t.event.trade_status==TradeStatus::Matched && t.event.buy_side);
 assert(t.event.trader_is_taker && !t.event.trader_is_maker);
 assert(t.event.taker_order_id.view()=="taker-123");
 assert(t.event.maker_order_count==2);
 assert(t.event.maker_orders[0].order_id.view()=="maker-a");
 assert(t.event.maker_orders[0].matched_amount.view()=="2.5");
 assert(!t.event.maker_orders[0].buy_side);
 assert(t.event.maker_orders[1].order_id.view()=="maker-b");

 auto envelope=decode(R"({"topic":"user","type":"trade","payload":{"id":"trade-env","taker_order_id":"taker-env","market":"0xCOND","asset_id":"TOKEN","side":"SELL","size":"1","price":"0.43","status":"CONFIRMED","trader_side":"MAKER","maker_orders":[]}})",121);
 assert(envelope.recognized && !envelope.invalid);
 assert(envelope.event.trade_status==TradeStatus::Confirmed);
 assert(envelope.event.trader_is_maker && !envelope.event.trader_is_taker);
 assert(envelope.event.taker_order_id.view()=="taker-env");

 auto p=decode("PONG",130);
 assert(p.recognized && !p.invalid && p.event.kind==EventKind::Pong);
 std::array<std::string_view,1> markets{"0xCOND"};
 auto s=subscription_json({"key-test","secret-test","pass-test"},markets);
 assert(s.find("\"apiKey\":\"key-test\"")!=std::string::npos);
 assert(s.find("\"markets\":[\"0xCOND\"]")!=std::string::npos);
 assert(s.find("\"type\":\"user\"")!=std::string::npos);
 assert(decode("not-json",140).invalid);
 assert(decode(R"({"event_type":"trade","id":"x","market":"m","asset_id":"a","side":"BUY","size":"1","price":"0.5","status":"MATCHED"})",141).invalid);
 assert(kHeartbeatIntervalMs==10000 && kHeartbeatRequest=="PING");
 std::cout << "PASS\n";
 return 0;
}
'''


def test_user_ws_codec_and_subscription() -> None:
    cxx = shutil.which("c++")
    assert cxx
    with tempfile.TemporaryDirectory() as td:
        path = Path(td)
        source = path / "main.cpp"
        binary = path / "user-ws-test"
        source.write_text(PROGRAM)
        subprocess.run(
            [
                cxx,
                "-std=c++20",
                "-O2",
                "-Wall",
                "-Wextra",
                "-Wpedantic",
                f"-I{ROOT / 'include'}",
                str(ROOT / "src/v7_user_ws.cpp"),
                str(ROOT / "src/boost_json.cpp"),
                str(source),
                "-o",
                str(binary),
            ],
            check=True,
            capture_output=True,
            text=True,
        )
        result = subprocess.run(
            [str(binary)], check=True, capture_output=True, text=True
        )
    assert result.stdout == "PASS\n"

File: tests/test_v7_user_ws.cpp
#include "pm/v7_user_ws.hpp"

#include <array>
#include <cassert>
#include <string_view>

using namespace pm::v7;

int main() {
    UserWsParser parser;
    std::array<UserWsEvent, 8> output{};
    const pm::fast::FeedReceiveStamp stamp{1'782'753'357'257LL, 9'000'000'000LL};

    constexpr std::string_view order = R"({
      "event_type":"order",
      "id":"0xorder",
      "market":"0xmarket",
      "asset_id":"123",
      "side":"BUY",
      "original_size":"10",
      "size_matched":"0",
      "price":"0.52",
      "type":"PLACEMENT",
      "status":"LIVE",
      "timestamp":"1782753357257"
    })";
    auto parsed = parser.parse(order, stamp, output);
    assert(parsed.invalid_frame == 0 && parsed.output_count == 1);
    assert(output[0].kind == UserWsEventKind::Order);
    assert(output[0].order_action == UserOrderAction::Placement);
    assert(output[0].order_status == UserOrderStatus::Live);
    assert(output[0].side == Side::Buy);
    assert(output[0].price_e4 == 5'200);
    assert(output[0].original_size_microunits == 10'000'000);
    assert(output[0].order_id.view() == "0xorder");
    assert(output[0].receive_monotonic_ns == stamp.monotonic_ns);

    constexpr std::string_view taker_trade = R"({
      "event_type":"trade",
      "id":"trade-1",
      "taker_order_id":"0xorder",
      "market":"0xmarket",
      "asset_id":"123",
      "side":"BUY",
      "size":"2.5",
      "price":"0.52",
      "status":"MATCHED",
      "timestamp":"1782753357258"
    })";
    parsed = parser.parse(taker_trade, stamp, output);
    assert(parsed.invalid_frame == 0 && parsed.output_count == 1);
    assert(output[0].kind == UserWsEventKind::Trade);
    assert(output[0].trade_status == UserTradeStatus::Matched);
    assert(output[0].trade_size_microunits == 2'500'000);
    assert(output[0].order_id.view() == "0xorder");
    assert(output[0].taker_order_id.view() == "0xorder");

    constexpr std::string_view maker_trade = R"({
      "event_type":"trade",
      "id":"trade-2",
      "taker_order_id":"foreign-taker",
      "market":"0xmarket",
      "asset_id":"123",
      "side":"SELL",
      "size":"5",
      "price":"0.51",
      "status":"MATCHED",
      "timestamp":"1782753357259",
      "maker_orders":[
        {"order_id":"our-maker","matched_amount":"1.25"},
        {"order_id":"other-maker","matched_amount":"3.75"}
      ]
    })";
    parsed = parser.parse(maker_trade, stamp, output);
    assert(parsed.invalid_frame == 0 && parsed.output_count == 3);
    assert(output[0].order_id.view() == "foreign-taker");
    assert(output[1].order_id.view() == "our-maker");
    assert(output[1].trade_size_microunits == 1'250'000);
    assert(output[2].order_id.view() == "other-maker");
    assert(output[2].trade_size_microunits == 3'750'000);

    constexpr std::string_view wrapped = R"({
      "type":"order",
      "payload":{
        "id":"wrapped-order",
        "market":"0xmarket",
        "asset_id":"123",
        "side":"SELL",
        "original_size":"3",
        "size_matched":"1",
        "price":"0.49",
        "type":"UPDATE",
        "status":"LIVE",
        "timestamp":1782753357260
      }
    })";
    parsed = parser.parse(wrapped, stamp, output);
    assert(parsed.invalid_frame == 0 && parsed.output_count == 1);
    assert(output[0].order_action == UserOrderAction::Update);
    assert(output[0].side == Side::Sell);
    assert(output[0].size_matched_microunits == 1'000'000);

    parsed = parser.parse("not-json", stamp, output);
    assert(parsed.invalid_frame != 0);
    return 0;
}
